### ISSUE TITLE
Prefer hand crafted Decode / Encode Typeclasses for nested Types

### ISSUE THREAD

**Issue by penland365** [alexarchambault/argonaut-shapeless#25]
** Note - It's possible I don't now how to use the library correctly.  Shapeless based libs are new enough to me that I'm not 100% sure this isn't currently possible

When dealing with deeply nested `Json`, (particularly 3rd party `Json`), it's not uncommon to have to have a `Json` field name be a non-valid parameter name, for example:
```json
{
  "points" :  {
    "for" : 2,
    "against" : 1
  }
}
```
In cases like this, we have to use custom `DecodeJson[Points]` typeclasses. This how bars us from using `implicitly[DecodeJson[A]]` for any type that composes the previously described `Points` type.  Example:
```json
{
  "name" : "Yankees",
  "location" : "New York City",
  "points" : {
    "for" : 2,
    "against" : 1
  }
}
```
Given the following code:
```scala
case class Points(pointsFor: Int, against: Int)
//decode typeclass provided
case class Team(name: String, location: String, points: Points)
implicit val decodeJson = implicitly[DecodeJson[Team]]
```
we fail to decode provided json with an error message of `cursor failed on --\ "pointsFor"`

I'm not even sure it's possible, but it would be nice if we could defer to any pre-defined typclasses for encoding / decoding.

**Comment by alexarchambault** [comment]
I was in the midst of addressing this. This will be definitely be fixed soon (with something like https://github.com/milessabin/shapeless/pull/427, in a separate project soon).

**Comment by 2chilled** [comment]
Upvote! Without this being fixed the library is useless for us at the moment.

**Comment by alexarchambault** [comment]
It's (just) fixed in snapshot (1.0.0-SNAPSHOT)! It depends on shapeless 2.3.0-SNAPSHOT though.

**Comment by 2chilled** [comment]
Great news, I'll try it right now

**Comment by 2chilled** [comment]
Unfortunately it does not work yet. Now I'm getting java.lang.StackOverflowError at implicitly[CodecJson[TypeWithNestedTypes]]. I verified the versions (shapeless snapshot too).

**Comment by alexarchambault** [comment]
This is fixed in master (version `1.0.0-SNAPSHOT`).
